src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -3854,9 +3854,9 @@ public JCTree.JCCompilationUnit parseCompilationUnit() {
             ListBuffer<JCTree> semiList = new ListBuffer<>();
             while (firstTypeDecl && mods == null && token.kind == SEMI) {
                 semiList.append(toP(F.at(token.pos).Skip()));
-                nextToken();
-                if (token.kind == EOF)
+                if (peekToken(EOF::equals))
                     break;
+                nextToken();
             }
```

Comment:
This might work, but it "feels" wrong (as in, a hack) to leave an extra semicolon around.

Yes, a package statement must be followed by a semicolon (line 3835) but after that can't you slurp up all the extra semicolons, and only report them if there is an import statement, and ignore them if there are no imports?

Reply:
I'd rather not mess with it because the existing code is already a customized (i.e., not a straightforward parse) so that it can provide a more helpful error message when duplicate semicolons are found after import statements. See [JDK-8027682](https://bugs.openjdk.org/browse/JDK-8027682) and PR #12448. Doing what you suggest would probably be more straightforward but would risk breaking that.

Reply:
> Doing what you suggest would probably be more straightforward

"More straightforward" is the right long term solution for the code.

> but would risk breaking that.

You now have tests to ensure you don't break that.

The previous round of fix was inadequate, for not handling the case of no import statements. It is reasonable to (re)consider the logic for that part of the code.  And, it is reasonable to ensure the set of test cases covers all places where multiple semicolons are bad.

Reply:
Isn't the logic just

* after a package declaration or import declaration, read and save excess semicolons
* before an import statement, check there are no saved excess semicolons
* (ignore any saved excess semicolons when you encounter the first type declaration (if any) or EOF)

Reply:
OK I get what you're saying now (I think) and this is easy to fix. See if you like f2de70cfb08 better.

Reply:
@jonathan-gibbons ... your thoughts?